### PR TITLE
Add Consensus Struct

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description
This PR adds a Consensus struct to the block chain. This controls what the consensus rules will be the values are called for reading

## Motivation and Context
See issue #641 

## How Has This Been Tested?
This does not break any current unit test. 
a Unit test has been added to confirm that the consensus struct changes rules

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/consensus.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+use crate::tari_amount::MicroTari;
+
+/// This is used to control all consensus values.
+pub struct ConsensusRules {
+    /// The min height maturity a coinbase utxo must have
+    pub coinbase_lock_height: u64,
+    /// The max range proof size that is allowed to be created
+    pub max_range_proof_range: usize,
+    /// This is emission schedule initial amount, the decay and the tail emission
+    pub emission_schedule: (MicroTari, f64, MicroTari),
```
Comment (mikethetike) [inline base_layer/core/src/consensus.rs]:
```suggestion
    pub emission_schedule: {initial_amount:MicroTari, decay:f64, tail:MicroTari},
```

File: base_layer/core/src/consensus.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        if cfg!(test) {
+            rules.coinbase_lock_height = 1;
+            rules.max_range_proof_range = 32;
+        }
+        rules
+    }
+}
+
+impl ConsensusRules {
+    pub fn current(&self) -> &Self {
```
Comment (mikethetike) [inline base_layer/core/src/consensus.rs]: Try using the singleton pattern here

```suggestion
    pub fn current() -> &Self {
```

File: base_layer/core/src/emission.rs [hunk 3]
```diff
@@ -46,10 +46,20 @@ impl EmissionSchedule {
...
     ///  * $$1-r$$ is the decay rate
     ///  * $$t$$ is the constant tail emission rate
-    pub fn new(initial: MicroTari, decay: f64, tail: MicroTari) -> EmissionSchedule {
+    pub fn new_with(initial: MicroTari, decay: f64, tail: MicroTari) -> EmissionSchedule {
         EmissionSchedule { initial, decay, tail }
     }
 
+    /// Creates a new emssion schedule instance from the conensus rules
+    pub fn new() -> EmissionSchedule {
+        let emission = CONSENSUS_RULES.current().emission_schedule.clone();
```
Comment (mikethetike) [inline base_layer/core/src/emission.rs]:
```suggestion
        let emission = ConsensusRules::current().emission_schedule;
```

**Comment by mikethetike** [comment]
Sorry, added some suggestions as per our offline discussion. Next time I'll put the discussion online here

File: base_layer/core/src/consensus.rs (hunk 1 above)
Comment (SWvheerden) [reply base_layer/core/src/consensus.rs]: We talked about this, I am of the opinion this struct should only contain values, no other objects like the emission schedule. 

The emission schedule has a running iterator in it, and which to me belongs more in the chain struct. 

The consensus struct should only be there for initialising objects and their parameters.

**Comment by SWvheerden** [comment]
I have my doubts about this route. This might increase test flakiness..
The problem is that single shared resource. It looks like integration tests have a different resource which is good as those tests require different rules.

**Pull request merged**